caput/pipeline.py

```diff
@@ -918,71 +1126,168 @@ def cacheable(self):
     # Pipeline Infrastructure
     # -----------------------
 
+    @property
+    def _pipeline_is_available(self):
+        """True if this task can be run."""
+        if not hasattr(self, "_pipeline_state"):
+            # This task hasn't been initialized
+            return False
+
+        if self._pipeline_state == "setup":
+            # True if all `requires` items have been provided
+            # This also returns True is `self._requires` is empty
+            return all(r is not None for r in self._requires)
+
+        if self._pipeline_state == "next":
+            # True if there is at least one input available
+            # in each input queue.
+            return bool(min((q.qsize() for q in self._in), default=0))
+
+        # Otherwise, this task is likely done and can be run to
+        # see if anything else happens
+        return True
+
+    @property
+    def priority(self):
+        """Return the priority associated with this task.
+
+        If the task is not yet initialized, return the base priority.
+
+        If the task in in state `setup`, return base priority
+        plus one if all `requires` items are stashed, otherwise base
+        priority.
+
+        If the task is in state `next`, priority is the total net consumption
+        of the task plus its base priority.
+
+        For example:
+        - A task which consumes 2 items, produces one, and can currently run
+        once will have priority (2 - 1) * 1 + base = 1 + base
+        - A task which does not consume anything but produces one item
+        will have priority (0 - 1) * 1 + base = -1 + base
+
+        In any other state, priority is just net consumption for one iteration
+        plus the base priority.
+        """
+        if not hasattr(self, "_pipeline_state"):
+            # This task hasn't been initialized
+            p = 0
+
+        elif self._pipeline_state == "setup":
+            # 1 if all requirements are available or no requirements,
+            # zero if requirements are needed but not available
+            p = int(all(r is not None for r in self._requires))
+
+        elif self._pipeline_state == "next":
+            # Calculate the total net consumption of the task
+            p = len(self._in_keys) - len(self._out_keys)
+            # How many times can the task run?
+            p *= min((q.qsize() for q in self._in), default=1)
+
+        else:
+            # If a task has passed the above states, it should be
+            # finished quickly so set a very high priority
+            p = 100
```

Comment:
It's not clear to me how the value 100 compares to the values computed in the previous block.  If the intent here is to make the priority in this case be always higher that what's calculated there, I'd make this significantly (orders of magnitude) larger, just to remove any hint of ambiguity.